File: MLP/OIM-SA/solvers/oim_dynamics.jl
module OIMDynamics

using LinearAlgebra
using Base.Threads
using Random
using Plots  # For visualization
using DifferentialEquations
using Statistics  # Add this for mean function

# Print detailed thread information at module load
println("\nJulia OIM solver initialized:")
println("- Total threads available: $(Threads.nthreads())")
println("- Thread IDs: $(collect(1:Threads.nthreads()))")
println("- Physical cores: $(div(Sys.CPU_THREADS, 2))")
println("- Total CPU threads: $(Sys.CPU_THREADS)\n")

export solve_batch_oim, solve_single_oim, test_oim_dynamics

"""
Compute OIM dynamics for a single state vector.
Optimized for maximum performance with SIMD, fastmath, and in-place operations.
"""
@fastmath function compute_dynamics!(du::Vector{Float64}, u::Vector{Float64}, 
                         J::Matrix{Float64}, h::Union{Vector{Float64},Nothing}, 
                         K_s::Union{Vector{Float64},Nothing})
    n = length(u)
    fill!(du, 0.0)
    
    # Compute oscillator coupling terms with optimized loops
    @inbounds for i in 1:n
        du_sum = 0.0
        @simd for j in 1:n
            if J[i,j] != 0  # Skip zero couplings
                du_sum -= J[i,j] * sin(u[i] - u[j])
            end
        end
        du[i] = du_sum
    end
    
    # External field (if provided)
    if !isnothing(h)
        @inbounds @simd for i in 1:n
            du[i] -= h[i] * sin(u[i])
        end
    end
    
    # SHIL sync (if provided)
    if !isnothing(K_s)
        @inbounds @simd for i in 1:n
            du[i] -= K_s[i] * sin(2 * u[i])
        end
    end
    
    return du
end

"""
Euler integration with fixed time step and history tracking.
"""
function euler_integration_with_history(u::Vector{Float64}, J::Matrix{Float64}, h::Union{Vector{Float64},Nothing}, K_s::Union{Vector{Float64},Nothing}, dt::Float64, duration::Float64)
    n_steps = Int(round(duration/dt)) + 1
    t_history = collect(range(0, duration, length=n_steps))
    u_history = zeros(length(u), n_steps)
    u_history[:, 1] = u
    t = 0.0
    step = 1
    du = similar(u)
    while t < duration
        compute_dynamics!(du, u, J, h, K_s)
        @. u += dt * du
        t += dt
        step += 1
        if step <= n_steps
            u_history[:, step] = u
        end
    end
    return t_history, u_history
end

"""
RK integration using DifferentialEquations.jl with adaptive time steps and history tracking.
"""
function rk_integration_with_history(u::Vector{Float64}, J::Matrix{Float64}, h::Union{Vector{Float64},Nothing}, K_s::Union{Vector{Float64},Nothing}, dt::Float64, duration::Float64)
    # Package parameters into a named tuple
    p = (J=J, h=h, K_s=K_s)
    
    # Pre-allocate the derivative vector
    du = similar(u)
    
    function f_de(du::Vector{Float64}, u::Vector{Float64}, p, t::Float64)
        compute_dynamics!(du, u, p.J, p.h, p.K_s)
    end
    
    tspan = (0.0, duration)
    prob = ODEProblem(f_de, u, tspan, p)
    
    # Option 1: BS3 with looser tolerances
    # sol = solve(prob, BS3(), dt=dt, abstol=4e-2, reltol=2e-2, saveat=dt, adaptive=true, dtmax=dt*10)
    
    # Option 2: BS3 with tighter tolerances
    sol = solve(prob, BS3(), dt=dt, abstol=1e-2, reltol=1e-2, saveat=dt, adaptive=true, dtmax=dt*5)
    
    # Option 3: Tsit5 with tighter tolerances
    # sol = solve(prob, Tsit5(), dt=dt, abstol=1e-2, reltol=1e-2, saveat=dt, adaptive=true, dtmax=dt*5)
    
    t_history = sol.t
    u_history = Array(sol)
    return t_history, u_history
end

"""
Fast Euler integration without saving history.
"""
function euler_integration(u::Vector{Float64}, J::Matrix{Float64}, h::Union{Vector{Float64},Nothing}, K_s::Union{Vector{Float64},Nothing}, dt::Float64, duration::Float64)
    t = 0.0
    du = similar(u)
    while t < duration
        compute_dynamics!(du, u, J, h, K_s)
        @. u += dt * du
        t += dt
    end
    return u
end

"""
Fast RK integration using DifferentialEquations.jl without saving history.
"""
function rk_integration(u::Vector{Float64}, J::Matrix{Float64}, h::Union{Vector{Float64},Nothing}, K_s::Union{Vector{Float64},Nothing}, dt::Float64, duration::Float64)
    p = (J=J, h=h, K_s=K_s)
    
    # Pre-allocate the derivative vector
    du = similar(u)
    
    function f_de(du::Vector{Float64}, u::Vector{Float64}, p, t::Float64)
        compute_dynamics!(du, u, p.J, p.h, p.K_s)
    end
    
    tspan = (0.0, duration)
    prob = ODEProblem(f_de, u, tspan, p)
    
    # Using BS3 with tighter tolerances and increased dtmax multiplier for better adaptivity
    sol = solve(prob, BS3(), dt=dt, abstol=1e-2, reltol=1e-2, save_everystep=false, save_start=false, adaptive=true, dtmax=dt*50)
    
    return sol.u[end]
end

"""
Solve OIM dynamics and track the full evolution.
"""
function solve_single_oim_with_history(J::Matrix{Float64}, 
                                     h::Union{Vector{Float64},Nothing},
                                     K_s::Union{Vector{Float64},Nothing};
                                     u0::Union{Vector{Float64},Nothing}=nothing,
                                     duration::Float64=20.0,
                                     dt::Float64=0.1,
                                     random_initialization::Bool=true,
                                     ode_solver::String="Euler")
    n = size(J, 1)

    # Initialize state
    if isnothing(u0)
        if random_initialization
            # Set fixed seed for reproducibility
            Random.seed!(42)
            u = rand(n) * 2π
        else
            u = fill(π/2, n)
        end
    else
        u = copy(u0)
    end
    
    if ode_solver == "Euler"
        t_history, u_history = euler_integration_with_history(u, J, h, K_s, dt, duration)
    elseif ode_solver == "RK"
        t_history, u_history = rk_integration_with_history(u, J, h, K_s, dt, duration)
    else
        error("Unknown ode_solver option: $ode_solver")
    end
    
    return u, t_history, u_history
end

"""
Solve OIM dynamics for a single problem using Euler integration.
"""
function solve_single_oim(J::Matrix{Float64}, 
                        h::Union{Vector{Float64},Nothing}=nothing,
                        K_s::Union{Vector{Float64},Nothing}=nothing;
                        u0::Union{Vector{Float64},Nothing}=nothing,
                        duration::Float64=20.0,
                        dt::Float64=0.1,
                        make_plot::Bool=false,
                        phase::String="Free",
                        output_start_idx::Int=120,
                        random_initialization::Bool=true,
                        target_idx=nothing,
                        ode_solver::String="Euler")
    
    if !make_plot
        # Fast path without plotting - use fast integration without saving history
        n = size(J, 1)
        
        # Initialize state
        if isnothing(u0)
            if random_initialization
                # Set fixed seed for reproducibility
                Random.seed!(42)
                u = rand(n) * 2π  # Random initial conditions between 0 and 2π
            else
                u = fill(π/2, n)  # Fixed initial conditions
            end
        else
            u = copy(u0)
        end
        if ode_solver == "Euler"
            u_final = euler_integration(u, J, h, K_s, dt, duration)
        elseif ode_solver == "RK"
            u_final = rk_integration(u, J, h, K_s, dt, duration)
        else
            error("Unknown ode_solver option: $ode_solver")
        end
        return u_final
    else
        # Path with plotting enabled - use solve_single_oim_with_history
        u, t_history, u_history = solve_single_oim_with_history(
            J, h, K_s,
            u0=u0,
            duration=duration,
            dt=dt,
            random_initialization=random_initialization,
            ode_solver=ode_solver
        )
        
        # Create plot using the existing plot_dynamics function
        plot_dynamics(t_history, u_history, 
                     phase=phase, 
                     output_start_idx=output_start_idx,
                     target_idx=target_idx)
        return u
    end
end

"""
Calculate phase velocities for the current state.
"""
function calculate_phase_velocities(u::Vector{Float64}, J::Matrix{Float64}, h::Union{Vector{Float64},Nothing}, K_s::Union{Vector{Float64},Nothing})
    du = similar(u)
    compute_dynamics!(du, u, J, h, K_s)
    return du
end

"""
Calculate various metrics to assess convergence of the dynamics.
"""
function get_convergence_metrics(u::Vector{Float64}, J::Matrix{Float64}, h::Union{Vector{Float64},Nothing}, K_s::Union{Vector{Float64},Nothing})
    velocities = calculate_phase_velocities(u, J, h, K_s)
    return Dict(
        "max_velocity" => maximum(abs.(velocities)),
        "mean_velocity" => mean(abs.(velocities)),
        "rms_velocity" => sqrt(mean(velocities.^2))
    )
end

"""
Solve OIM dynamics for a batch of problems in parallel.
Now returns both final states and convergence metrics.
"""
function solve_batch_oim(J_batch::Vector{Matrix{Float64}},
                        h_batch::Union{Vector{Vector{Float64}},Nothing}=nothing,
                        K_s_batch::Union{Vector{Vector{Float64}},Nothing}=nothing;
                        u0_batch::Union{Vector{Vector{Float64}},Nothing}=nothing,
                        duration::Float64=20.0,
                        dt::Float64=0.1,
                        n_procs::Union{Int,Nothing}=nothing,
                        make_plot::Bool=false,
                        phase::String="Free",
                        output_start_idx::Int=120,
                        random_initialization::Bool=true,
                        target_idx::Union{Int,Nothing}=nothing,
                        ode_solver::String="Euler")
    
    n_problems = length(J_batch)
    results = Vector{Vector{Float64}}(undef, n_problems)
    metrics = Vector{Dict{String,Float64}}(undef, n_problems)
    
    # Determine number of threads to use
    max_threads = Threads.nthreads()
    n_threads = isnothing(n_procs) ? max_threads : min(n_procs, max_threads)
    
    # Process problems in chunks based on number of threads
    chunk_size = ceil(Int, n_problems / n_threads)
    
    # Process all problems in parallel using threads
    @threads for chunk in 1:n_threads
        start_idx = (chunk - 1) * chunk_size + 1
        end_idx = min(chunk * chunk_size, n_problems)
        
        for i in start_idx:end_idx
            h_i = isnothing(h_batch) ? nothing : h_batch[i]
            K_s_i = isnothing(K_s_batch) ? nothing : K_s_batch[i]
            u0_i = isnothing(u0_batch) ? nothing : u0_batch[i]
            
            # Solve single problem
            results[i] = solve_single_oim(
                J_batch[i],
                h_i,
                K_s_i,
                u0=u0_i,
                duration=duration,
                dt=dt,
                make_plot=make_plot && i == 1, # Only plot the first problem in debug mode
                phase=phase,
                output_start_idx=output_start_idx,
                random_initialization=random_initialization,
                target_idx=target_idx,
                ode_solver=ode_solver
            )
            
            # Calculate convergence metrics
            metrics[i] = get_convergence_metrics(results[i], J_batch[i], h_i, K_s_i)
        end
    end
    
    return results, metrics
end

"""
Plot phase evolution over time and save to file.
"""
function plot_dynamics(t::Vector{Float64}, phases::Matrix{Float64}; 
                      title::String="Phase Evolution",
                      phase::String="Free",
                      savepath::String="oim_dynamics_plot.png",
                      output_start_idx::Int=120,  # Default for 120 hidden, 10 output
                      target_idx=nothing)  # Accept any type and convert inside
    n_phases = size(phases, 1)
    p = plot()  # Create empty plot
    
    # Convert target_idx to Int if it's not nothing
    target_idx_int = nothing
    if !isnothing(target_idx)
        target_idx_int = convert(Int, target_idx)
    end
    
    # Plot hidden layer phases with default colors
    plot!(p, t, phases[1:output_start_idx,:]', 
         label=nothing,
         alpha=0.7,
         linewidth=1)  # Thinner lines for hidden neurons
    
    # Plot output layer phases with distinct styles
    if output_start_idx < n_phases
        # Define a set of vibrant colors for output neurons
        output_colors = [:black, :red, :blue, :green, :purple, :orange, :cyan, :magenta, :brown, :gold]
        line_styles = [:dash]
        
        for (i, output_idx) in enumerate(output_start_idx+1:n_phases)
            color_idx = mod1(i, length(output_colors))
            style_idx = mod1(i, length(line_styles))
            
            # If this is the target neuron, give it a special color and label
            if !isnothing(target_idx_int) && (i-1) == target_idx_int
                plot!(p, t, phases[output_idx:output_idx,:]',
                     label="Target (Class $target_idx_int)",
                     alpha=1.0,
                     linewidth=3,
                     color=:lime,  # Neon green for target
                     linestyle=:solid)
            else
                plot!(p, t, phases[output_idx:output_idx,:]',
                     label=nothing,
                     alpha=1.0,
                     linewidth=3,
                     color=output_colors[color_idx],
                     linestyle=line_styles[style_idx])
            end
        end
    end
    
    # Set plot attributes
    plot!(p, title="$phase Phase Evolution",
         xlabel="Time",
         ylabel="Phase",
         legend=:right)  # Show legend inside on the right
    
    # Modify filename to include phase
    base, ext = splitext(savepath)
    savepath = "$(base)_$(lowercase(phase))$(ext)"
    savefig(p, savepath)
    return p
end

"""
Test the OIM dynamics solver on a simple example.
"""
function test_oim_dynamics(; n_spins::Int=10, plot_result::Bool=true)
    println("\nRunning OIM dynamics test:")
    println("- Number of spins: $n_spins")
    
    # Create a simple coupling matrix 
    J = zeros(n_spins, n_spins)
    for i in 1:n_spins-1
        J[i,i+1] = J[i+1,i] = -1.0
    end
    println("- Created nearest-neighbor coupling matrix")
    
    # Create simple external field and SHIL sync
    h = fill(0.0, n_spins)
    K_s = fill(10.0, n_spins)
    println("- Created uniform external field and SHIL sync vectors")
    
    # Random initial conditions
    u0 = rand(n_spins) * 2π
    println("- Generated random initial conditions")
    
    # Solve with history tracking
    println("- Solving dynamics...")
    t, history = solve_single_oim_with_history(
        J, h, K_s,
        u0=u0,
        duration=20.0,
        dt=0.1
    )
    
    # Print results
    println("- Initial phases: ", round.(u0, digits=3))
    println("- Final phases: ", round.(history[:,end], digits=3))
    
    # Plot if requested
    if plot_result
        println("- Plotting results...")
        plot_dynamics(t, history, title="Test Case: $n_spins-Spin Chain")
    end
    
    return t, history
end

end # module 

File: MLP/OIM-SA/solvers/oim_dynamics.py
import os

# Set OpenMP environment variable to avoid fork issues
# These must be set before importing numpy or other OpenMP-using libraries
os.environ['OPENBLAS_NUM_THREADS'] = '1'
os.environ['MKL_NUM_THREADS'] = '1'
os.environ['OMP_NUM_THREADS'] = '1'
os.environ['NUMEXPR_NUM_THREADS'] = '1'

import numpy as np
import torch
from scipy.integrate import solve_ivp
from numba import jit, prange
import warnings
from typing import Optional, Union, List, Tuple
import math
import matplotlib.pyplot as plt
import multiprocessing as mp
from functools import partial
import time

class OIMDynamics:
    """
    OIM dynamics solver with fixed coupling parameters.
    Supports both CPU (Numba) and GPU (PyTorch) acceleration.
    """
    def __init__(self, 
                 J: Union[np.ndarray, torch.Tensor], 
                 h: Optional[Union[np.ndarray, torch.Tensor]] = None,
                 K_s: Optional[Union[np.ndarray, torch.Tensor]] = None,
                 use_gpu: bool = False):
        """
        Initialize OIM dynamics solver with fixed parameters.
        
        Args:
            J: Coupling matrix (fixed in time)
            h: External field vector (fixed in time)
            K_s: SHIL sync vector (fixed in time)
            use_gpu: Whether to use GPU acceleration
        """
        self.use_gpu = use_gpu and torch.cuda.is_available()
        
        # Move problem parameters to GPU if requested
        if self.use_gpu:
            self.device = torch.device('cuda')
            self.J = torch.as_tensor(J, device=self.device, dtype=torch.float32)
            self.h = torch.as_tensor(h, device=self.device, dtype=torch.float32) if h is not None else None
            self.K_s = torch.as_tensor(K_s, device=self.device, dtype=torch.float32) if K_s is not None else None
        else:
            self.J = np.asarray(J, dtype=np.float32)
            self.h = np.asarray(h, dtype=np.float32) if h is not None else None
            self.K_s = np.asarray(K_s, dtype=np.float32) if K_s is not None else None
            
        self.n_spins = len(J)
    
    @staticmethod
    @jit(nopython=True, parallel=True)
    def _dynamics_numba(u: np.ndarray, J: np.ndarray, h: Optional[np.ndarray], 
                       K_s: Optional[np.ndarray]) -> np.ndarray:
        """Numba-accelerated OIM dynamics with fixed parameters."""
        n = len(u)
        du = np.zeros_like(u)
        
        # Oscillator coupling - parallelized over spins
        for i in prange(n):
            for j in range(n):
                if J[i,j] != 0:  # Skip zero couplings
                    phase_diff = u[i] - u[j]
                    du[i] -= J[i,j] * np.sin(phase_diff)
        
        # External field
        if h is not None:
            du -= h * np.sin(u)
            
        # SHIL sync
        if K_s is not None:
            du -= K_s * np.sin(2 * u)
        
        return du

    def _dynamics_gpu(self, u: torch.Tensor) -> torch.Tensor:
        """GPU-accelerated OIM dynamics using PyTorch."""
        # Reshape for broadcasting
        u_i = u.view(-1, 1) # column vector
        u_j = u.view(1, -1) # row vector

        # Compute all phase differences at once (i.e. u_i - u_j is a matrix)
        phase_diffs = u_i - u_j
        
        # Oscillator coupling using matrix operations
        du = -torch.sum(self.J * torch.sin(phase_diffs), dim=1)
        
        # External field
        if self.h is not None:
            du -= self.h * torch.sin(u)
            
        # SHIL sync
        if self.K_s is not None:
            du -= self.K_s * torch.sin(2 * u)
        
        return du

    def dynamics(self, t: float, u: Union[np.ndarray, torch.Tensor]) -> np.ndarray:
        """Main dynamics function that routes to appropriate implementation."""
        if self.use_gpu:
            u_tensor = torch.as_tensor(u, device=self.device, dtype=torch.float32)
            return self._dynamics_gpu(u_tensor).cpu().numpy()
        else:
            u_array = np.asarray(u, dtype=np.float32)
            return self._dynamics_numba(u_array, self.J, self.h, self.K_s)

    def solve(self, 
             u0: Union[np.ndarray, torch.Tensor], 
             duration: float, 
             dt: float,
             method: str = 'RK45',
             noise: bool = False,
             noise_strength: float = 0.0,
             plot_dynamics: bool = False) -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray, np.ndarray]]:
        """
        Solve OIM dynamics.
        
        Args:
            u0: Initial state
            duration: Total simulation time
            dt: Time step
            method: Integration method ('RK45' or 'Euler')
            noise: Whether to add noise
            noise_strength: Strength of noise term
            plot_dynamics: Whether to plot phase evolution
            
        Returns:
            If plot_dynamics=False: Final state
            If plot_dynamics=True: Tuple of (final state, time points, phase evolution)
        """
        if method == 'Euler':
            return self._solve_euler(u0, duration, dt, noise, noise_strength, plot_dynamics)
        else:
            with warnings.catch_warnings():
                warnings.simplefilter("ignore")
                sol = solve_ivp(
                    self.dynamics,
                    (0, duration),
                    np.asarray(u0),
                    method=method,
                    t_eval=np.arange(0, duration, dt),
                    rtol=1e-6,
                    atol=1e-6
                )
                
            if plot_dynamics:
                self._plot_dynamics(sol.t, sol.y)
                return sol.y[:,-1], sol.t, sol.y
            
            return sol.y[:,-1]

    def _solve_euler(self, 
                    u0: Union[np.ndarray, torch.Tensor], 
                    duration: float, 
                    dt: float,
                    noise: bool = False,
                    noise_strength: float = 0.0,
                    plot_dynamics: bool = False) -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray, np.ndarray]]:
        """Simple Euler integration with optional noise and plotting."""
        steps = int(duration / dt)
        t = np.arange(0, duration, dt)
        u = np.asarray(u0, dtype=np.float32)
        
        if plot_dynamics:
            u_history = np.zeros((len(u), len(t)))
            u_history[:,0] = u
        
        for i, t_i in enumerate(t[:-1]): # t[:-1] means we don't include the last time point
            du = self.dynamics(t_i, u)
            if noise:
                du += noise_strength * np.random.randn(self.n_spins).astype(np.float32)
            u += dt * du
            
            if plot_dynamics:
                u_history[:,i+1] = u # we calculate u at the next time point and store it
        
        if plot_dynamics:
            self._plot_dynamics(t, u_history)
            return u, t, u_history
            
        return u

    def _plot_dynamics(self, t: np.ndarray, phases: np.ndarray):
        """Plot phase evolution over time for all neurons and save to file."""
        plt.figure(figsize=(12, 6))
        
        # Plot all phases with same style
        for i in range(phases.shape[0]):
            plt.plot(t, phases[i], '-', alpha=0.7)
        
        plt.title('Phase Evolution Over Time')
        plt.xlabel('Time')
        plt.ylabel('Phase')
        plt.grid(True, alpha=0.3)
        plt.tight_layout()
        plt.savefig('oim_dynamics.png', dpi=300, bbox_inches='tight')
        plt.close()

def solve_single(args):
    """Helper function to solve a single OIM problem."""
    J, h, K_s, u0, params = args
    solver = OIMDynamics(J, h, K_s, use_gpu=params['use_gpu'])
    if u0 is None:
        u0 = np.full(len(J), np.pi/2, dtype=np.float32) # Note if u0 is None then we use the free phase initial conditions ie. u0 = [pi/2, pi/2, ...]
        
    final_state = solver.solve(
        u0, 
        params['duration'], 
        params['dt'], 
        params['method'], 
        params['noise'], 
        params['noise_strength'],
        plot_dynamics=params['plot_dynamics']
    )
    
    if params['plot_dynamics']:
        final_state = final_state[0]
        
    return final_state

def batch_solve_oim(J_batch: List[np.ndarray],
                   h_batch: Optional[List[np.ndarray]] = None,
                   K_s_batch: Optional[List[np.ndarray]] = None,
                   u0_batch: Optional[List[np.ndarray]] = None,
                   duration: float = 20.0,
                   dt: float = 0.1,
                   use_gpu: bool = False,
                   method: str = 'RK45',
                   noise: bool = False,
                   noise_strength: float = 0.0,
                   batch_size: int = 20,
                   plot_dynamics: bool = False,
                   n_procs: Optional[int] = None) -> List[np.ndarray]:
    """
    Solve OIM dynamics for a batch of problems in parallel.
    Processes in smaller chunks to manage memory.
    
    Args:
        J_batch: List of coupling matrices
        h_batch: List of external fields (optional)
        K_s_batch: List of SHIL sync vectors (optional)
        u0_batch: List of initial states (optional)
        duration: Total simulation time
        dt: Time step
        use_gpu: Whether to use GPU acceleration
        method: Integration method
        noise: Whether to add noise
        noise_strength: Strength of noise term
        batch_size: Size of chunks to process
        plot_dynamics: Whether to plot phase evolution for each problem
        n_procs: Number of processors to use for parallel processing. If None, uses (cpu_count - 1)
        
    Returns:
        List of final states
    """
    n_problems = len(J_batch)
    results = []
    
    # Determine number of processors to use
    if n_procs is None:
        n_procs = max(1, mp.cpu_count() - 1)
    
    # Shared parameters dictionary
    params = {
        'duration': duration,
        'dt': dt,
        'use_gpu': use_gpu,
        'method': method,
        'noise': noise,
        'noise_strength': noise_strength,
        'plot_dynamics': plot_dynamics
    }
    
    # Process in batches
    for i in range(0, n_problems, batch_size):
        batch_end = min(i + batch_size, n_problems)
        batch_J = J_batch[i:batch_end]
        batch_h = h_batch[i:batch_end] if h_batch is not None else [None] * (batch_end - i)
        batch_K_s = K_s_batch[i:batch_end] if K_s_batch is not None else [None] * (batch_end - i)
        batch_u0 = u0_batch[i:batch_end] if u0_batch is not None else [None] * (batch_end - i)
        
        # Create argument tuples for parallel processing
        args_list = [(J, h, K_s, u0, params) for J, h, K_s, u0 in zip(batch_J, batch_h, batch_K_s, batch_u0)]
        
        # Process batch in parallel
        with mp.Pool(n_procs) as pool:
            batch_results = pool.map(solve_single, args_list)
            results.extend(batch_results)
        
        # Force cleanup
        if use_gpu:
            torch.cuda.empty_cache()
    
    return results

def test_oim_dynamics():
    """Run a test of OIM dynamics with a simple example problem."""
    # Create a small test problem
    n_spins = 5
    
    # Create a simple coupling matrix with nearest-neighbor interactions
    J = np.zeros((n_spins, n_spins))
    for i in range(n_spins-1):
        coupling = np.random.choice([-1.0, 1.0])  # Random ±1 coupling
        J[i,i+1] = J[i+1,i] = coupling  # Symmetric coupling between neighbors
        
    # Create a simple external field
    h = np.ones(n_spins) * 0.0
    
    # Create SHIL sync vector
    K_s = np.ones(n_spins) * 0.0
    
    # Create solver
    solver = OIMDynamics(J, h, K_s)
    
    # Initial condition
    u0 = np.random.uniform(0, 2*np.pi, n_spins)
    
    # Solve with plotting
    final_state = solver.solve(
        u0=u0,
        duration=20.0,
        dt=0.1,
        method='Euler',
        noise=False,
        noise_strength=0.0,
        plot_dynamics=False
    )
    
    print("Initial phases:", u0)
    print("Final phases:", final_state)

def test_batch_oim_dynamics():
    """Test parallel batch processing with multiple example problems."""
    print("\nTesting batch OIM dynamics processing...")
    
    # Create multiple test problems
    n_problems = 12  # Number of test problems
    n_spins = 5     # Size of each problem
    n_procs = 4     # Number of processors to use
    
    print(f"Using {n_procs} processors for parallel processing")
    
    # Initialize batch lists
    J_batch = []
    h_batch = []
    K_s_batch = []
    u0_batch = []
    
    # Create test problems
    for _ in range(n_problems):
        # Create coupling matrix with random nearest-neighbor interactions
        J = np.zeros((n_spins, n_spins))
        for i in range(n_spins-1):
            coupling = np.random.choice([-1.0, 1.0])
            J[i,i+1] = J[i+1,i] = coupling
        J_batch.append(J)
        
        # Create random external fields
        h = np.random.uniform(-0.5, 0.5, n_spins)
        h_batch.append(h)
        
        # Create random SHIL sync vectors
        K_s = np.random.uniform(0, 0.2, n_spins)
        K_s_batch.append(K_s)
        
        # Create random initial conditions
        u0 = np.random.uniform(0, 2*np.pi, n_spins)
        u0_batch.append(u0)
    
    # Test different batch sizes
    batch_sizes = [6, 12]
    
    for batch_size in batch_sizes:
        print(f"\nTesting with batch_size = {batch_size}")
        
        # Time the parallel execution
        start_time = time.time()
        results = batch_solve_oim(
            J_batch=J_batch,
            h_batch=h_batch,
            K_s_batch=K_s_batch,
            u0_batch=u0_batch,
            duration=10.0,
            dt=0.1,
            method='Euler',
            noise=False,
            noise_strength=0.0,
            batch_size=batch_size,
            n_procs=n_procs
        )
        end_time = time.time()
        
        print(f"Number of problems solved: {len(results)}")
        print(f"Time taken: {end_time - start_time:.2f} seconds")
        print(f"Average time per problem: {(end_time - start_time) / n_problems:.2f} seconds")
        
        # Print sample results
        print("\nSample results (first 2 problems):")
        for i in range(min(2, len(results))):
            print(f"Problem {i+1} final phases:", results[i])

def test_deterministic_behavior():
    """Test that identical inputs produce identical outputs across multiple runs."""
    print("\nTesting deterministic behavior...")
    
    # Create a fixed test problem
    n_spins = 5
    
    # Fixed coupling matrix
    J = np.zeros((n_spins, n_spins))
    for i in range(n_spins-1):
        J[i,i+1] = J[i+1,i] = 1.0  # All ferromagnetic couplings
    
    # Fixed external field and SHIL sync
    h = np.ones(n_spins) * 0.5
    K_s = np.ones(n_spins) * 0.1
    
    # Fixed initial condition
    u0 = np.array([0.1, 0.2, 0.3, 0.4, 0.5])
    
    # Parameters for multiple trials
    n_trials = 5
    results = []
    
    print("Running multiple trials with identical inputs...")
    print(f"Initial phases: {u0}")
    
    # Run multiple trials
    for trial in range(n_trials):
        solver = OIMDynamics(J, h, K_s)
        final_state = solver.solve(
            u0=u0,
            duration=10.0,
            dt=0.1,
            method='Euler',
            noise=False,
            noise_strength=0.0,
            plot_dynamics=False
        )
        results.append(final_state)
        print(f"Trial {trial + 1} final phases: {final_state}")
    
    # Check if all results are identical
    reference = results[0]
    all_identical = all(np.allclose(result, reference, rtol=1e-5) for result in results[1:])
    
    if all_identical:
        print("\nTest PASSED: All trials produced identical results")
    else:
        print("\nTest FAILED: Trials produced different results")
        # Print the maximum difference between any two results
        max_diff = max(np.max(np.abs(result - reference)) for result in results[1:])
        print(f"Maximum difference between trials: {max_diff}")

    # Now test with parallel batch processing
    print("\nTesting deterministic behavior with parallel processing...")
    
    # Create batch with identical problems
    n_problems = 4
    J_batch = [J.copy() for _ in range(n_problems)]
    h_batch = [h.copy() for _ in range(n_problems)]
    K_s_batch = [K_s.copy() for _ in range(n_problems)]
    u0_batch = [u0.copy() for _ in range(n_problems)]
    
    # Run batch processing multiple times
    batch_results = []
    for trial in range(3):
        results = batch_solve_oim(
            J_batch=J_batch,
            h_batch=h_batch,
            K_s_batch=K_s_batch,
            u0_batch=u0_batch,
            duration=10.0,
            dt=0.1,
            method='Euler',
            noise=False,
            noise_strength=0.0,
            batch_size=2,
            n_procs=2
        )
        batch_results.append(results)
        print(f"\nBatch trial {trial + 1} results:")
        for i, result in enumerate(results):
            print(f"Problem {i + 1}: {result}")
    
    # Check if all batch results are identical
    batch_reference = batch_results[0]
    batch_identical = all(
        all(np.allclose(result, ref, rtol=1e-5) 
            for result, ref in zip(trial_results, batch_reference))
        for trial_results in batch_results[1:]
    )
    
    if batch_identical:
        print("\nBatch test PASSED: All parallel trials produced identical results")
    else:
        print("\nBatch test FAILED: Parallel trials produced different results")
        # Print the maximum difference between any two batch results
        max_batch_diff = max(
            max(np.max(np.abs(result - ref)) 
                for result, ref in zip(trial_results, batch_reference))
            for trial_results in batch_results[1:]
        )
        print(f"Maximum difference between parallel trials: {max_batch_diff}")

if __name__ == "__main__":
    import multiprocessing as mp

    # Switch to the 'spawn' start method at the very beginning:
    mp.set_start_method('spawn', force=True)

    test_oim_dynamics()
    test_batch_oim_dynamics()
    test_deterministic_behavior()